Add Audio File Support

MuchQuak opened the pull request:
# Issues #1389 #1554 #1586 #1587 #1588 #1599 #1600 #1601 #1692 #1693 #1694 #1696 #1697 #1698 #1699 #1673 #1711

This is more of a draft pr currently to get the review process started because looking at the file changed count it is gonna take a while to get it though.

# Summary
This is a total overhaul of the image systems to be generalized to any media with a current focus on audio but with video support in mind. This is a very large pr because it is a very large overhaul of deeply embedded systems

## Design
The general idea was to make a static class that was easy to use for the manager classes and other misc classes. Having it extend manager would be a hassle if their was a manager being injected into another manager or dealing with more polymorphic managers when executing a variety of post actions. Since it was static some utility classes needed to be made for cleaning input and output as well as database connections.

## Backwards Compatibility Issues
I have gone back and retroactively fixed most of the backwards compatibility issues that were introduced prior to our discussion on the topic. There might be some I missed though.

## Classes
- `Media` : Handles crud operations on media table and its related tables
- `MediaException` : For media specific errors commonly used to bubble up status messages for the ui
- `StorageStrategy`: Class aimed at getting our file storage logic under a single api that could be generalized for remote storage options.
- `LocalStorage`: Extends StorageStrategy as is used for storing, deleting, renaming, and remapping files on local storage
- `Sanitize`: Sanitizing utility that moves all the Manager sanitation logic into a static class with a simple api, only `out` for output and `in` for input sanitation. This was talked about a while ago and I figured I would take a crack at implementing it since it was relevant to work I was already doing.
- `Database`: Allows for composable database connection access for static classes without need for inheritance.

A lot of these classes are in the media.php file and I think that make sense for now, but I would be happy to split them into separate files if their is a strong opposition to how things are organized.

## Language Files
There is a lot of language file changes that Katie and others have worked on reviewing I would ignore them when doing the code review. 

## Deprecating
### OccurrenceEditorImages
Shouldn't be used for media upload and edits anymore
Still Used:
- `webservices/imageuploader.php` - Not used anywhere else could this be removed?
- `imageoccursubmit.php` - No plans on updating this to use `Media` class but could if desired.

### ImageDetailManager
Shouldn't be used moving forward
Still used:
- `imgdetails.php` - only for `getDomain` function could easily get removed if the function was made available via utilities.
- `ActionManager` - Not really sure what this is did not touch.

### Not Deprecated but could be in the future
- `ImageShared.php` - Touched to many other parts of the code base so didn't implement `Media` class in place of it to save time and scope creep.

### How to Test
Media could be image or audio both should be tested to make sure system work with both
- Occurrence Editor
  - Upload from media
  - Upload from remote
  - Upload Larger file than max upload size (should give you message)
  - Update Metadata
  - Disassociate media
  - Remap media
  - Delete media
- Taxon Profile
  - Check that taxa profile page has audio displayed in media gallery as a speaker icon. Should not be able to play audio from taxon profile as of now.
  - Edit Media Metadata from taxon profile admin page (Should take you media detail page)
- Media Detail
  - Update
  - Unlink from taxon
  - Delete
- Media Search
  - Search for only audio
  - Search for only images
- Occurrence Profile
  - Uploaded Audio should be displayed on occurrence page and be playable
- Upload
  - Audio can be uploaded via collprofile -> Import/Update Specimen Records -> Extended Data Import 
- Download 
  - any download with media options selected. Make sure exported links are correct for hosted media and external media

# Pull Request Checklist:

# Pre-Approval

- [x] There is a description section in the pull request that details what the proposed changes do. It can be very brief if need be, but it ought to exist.
- [x] Hotfixes should be branched off of the `master` branch and **squash and merged** back into the `master` branch.
- [x] Features and backlog bugs should be merged into the `Development` branch, **NOT** `master`
- [x] All new text is preferably internationalized (i.e., no end-user-visible text is hard-coded on the PHP pages), and the [spreadsheet tracking internationalizations](https://docs.google.com/spreadsheets/d/133fps9w2pUCEjUA6IGCcQotk7dn9KvepMXJ2IWUZsE8/edit?usp=sharing) has been updated either with a new row or with checkmarks to existing rows.
- [x] There are no linter errors
- [x] New features have responsive design (i.e., look aesthetically pleasing both full screen and with small or mobile screens)
- [x] [Symbiota coding standards](https://docs.google.com/document/d/1-FwCZP5Zu4f-bPwsKeVVsZErytALOJyA2szjbfSUjmc/edit?usp=sharing) have been followed
- [x] If any files have been reformatted (e.g., by an autoformatter), the reformat is its own, separate commit in the PR
- [x] Comment which GitHub issue(s), if any does this PR address
- [x] If this PR makes any changes that would require additional configuration of any Symbiota portals outside of the files tracked in this repository, make sure that those changes are detailed in [this document](https://docs.google.com/document/d/1T7xbXEf2bjjm-PMrlXpUBa69aTMAIROPXVqJqa2ow_I/edit?usp=sharing).

# Post-Approval

- [x] It is the code author's responsibility to merge their own pull request after it has been approved
- [x] If this PR represents a merge into the `Development` branch, remember to use the **squash & merge** option
- [x] If this PR represents a merge from the `Development` branch into the master branch, remember to use the **merge** option
- [x] If this PR represents a hotfix into the `master` branch, a subsequent PR from `master` into `Development` should be made **merge** option (i.e., no squash).
- [x] If the dev team has agreed that this PR represents the last PR going into the `Development` branch before a tagged release (i.e., before an imminent merge into the master branch), make sure to notify the team and [lock the `Development` branch](https://github.com/BioKIC/Symbiota/settings/branches) to prevent accidental merges while QA takes place. Follow the release protocol [here](https://github.com/BioKIC/Symbiota/blob/master/docs/release-protocol.md).
- [x] Don't forget to delete your feature branch upon merge. Ignore this step as required.

Thanks for contributing and keeping it clean!

MuchQuak committed:
initial pass at changing image table to media table

there may be quite a few issues with regards to photographer and photographeruid being passed in from a postArray before things will work properly

MuchQuak committed:
missed because of unsaved file

MuchQuak committed:
Merge branch 'Development' into image-to-media

MuchQuak committed:
taxon profile using new img id

MuchQuak committed:
removing duplicate file from saving

MuchQuak committed:
updating branch with Development

MuchQuak committed:
changing imgid to media_id

MuchQuak committed:
cleanup errors on using new media table on occurrence editor and individual

MuchQuak committed:
fixing image search to work with new media table

MuchQuak committed:
cleaning up last references to images table

MuchQuak committed:
clean of use of creator instead of photographer data

MuchQuak committed:
fix for insert images

MuchQuak committed:
Fixing error with occurrence editor and photographer

MuchQuak committed:
pulling dev updates

MuchQuak committed:
fix image delete functionality

MuchQuak committed:
editor fix

MuchQuak committed:
add new Media and Sanitize Class. Also a database trait to help break away from manager

MuchQuak committed:
update image tab to use media class

MuchQuak committed:
Update lang tags to be media related on imagetabs and occurrence editor

MuchQuak committed:
add hokey stop gap to only get width and height of inserted media if png or jpeg should be any image

MuchQuak committed:
commenting out test echos for connection usage

MuchQuak committed:
remove testing db pool class

MuchQuak committed:
fixing media classes isRemapplable

MuchQuak committed:
remove limiter on getImageMap for media type

MuchQuak committed:
fix Santize to use iterate over $val instead of $arr

MuchQuak committed:
update santize to return null if string can be identified

This is to address the problem that integers coming from forms without a
default value would be evalutate as strings. So the default is now null
and strings that are empty that need to be stored that as empty strings
instead of null should be handled accordingly.

MuchQuak committed:
Added UploadStrategy, SymbiotaUploadStrategy classes and addMedia function to media

MuchQuak committed:
fix path construction in the symbiota upload strategy

MuchQuak committed:
Merge branch 'Development' into image-to-media

MuchQuak committed:
add rudamentry duplicate checking and recover

MuchQuak committed:
Add doc comments for Upload Strategy and Media. Refine api interactions for media and Upload Strategy

MuchQuak committed:
add remote media resource adding to form

MuchQuak committed:
Progress on Media::add edge cases surrounding images

MuchQuak committed:
implement image resize functions for gd and imagick

MuchQuak committed:
port system call to new media class imagick support

MuchQuak committed:
Edit db trait to keep write and read connections

MuchQuak committed:
update imagetab photographer to use creator

MuchQuak committed:
clean up and modification of UploadStrategy api and image functions

MuchQuak committed:
fixing some logic bugs in add media

MuchQuak committed:
removing debug echo

MuchQuak committed:
implement delete for Media class

MuchQuak committed:
Fetch updates from Development

MuchQuak committed:
add remove files flag to media::delete

MuchQuak committed:
use media delete flag from post request

MuchQuak committed:
implement media tag for Media class

MuchQuak committed:
remap media progress

MuchQuak committed:
fixed remapping media to correctly update the file path of the new occurrence

MuchQuak committed:
remove debug info on remap

MuchQuak committed:
rework remap media to not try and remap files than have already been added to the move stack

MuchQuak committed:
implement media diassociating

MuchQuak committed:
form errors for imgdetails

MuchQuak committed:
update imgdetails.php to work with any media

MuchQuak committed:
imagetab lang english lang update

MuchQuak committed:
get media by taxa id

MuchQuak committed:
Fetching Updates from Development

MuchQuak committed:
pullin in 3.1 final changes

MuchQuak committed:
fixing imgDetails submit and transfer image form

MuchQuak committed:
add audio files to taxon profile page

MuchQuak committed:
basic render function for media_items

MuchQuak committed:
fixing media_item to just no op media_item instead of erro

MuchQuak committed:
add media:render

MuchQuak committed:
add media to occurrence individual

MuchQuak committed:
add tpeditor support for images

MuchQuak committed:
get tpeditor working with Media class. Make adjusts to UploadStrategy to take path instead of vars to create path

MuchQuak committed:
old typo on creatorDisplay

MuchQuak committed:
adjust occurrence media fetch to put null sortoccurrence on bottom and low sortoccurrenc on top

MuchQuak committed:
imagesearch not shows audio results

MuchQuak committed:
add options to image search to query per media type

MuchQuak committed:
removing debug errorlog

MuchQuak committed:
fix image lib media_type param null check

MuchQuak committed:
updated translations for imaglib/search

MuchQuak committed:
add english media_types for imglib/search

MuchQuak committed:
image -> media translations taxa index

MuchQuak committed:
update missing translations for taxa/index

MuchQuak committed:
tp editor image -> media translations

MuchQuak committed:
tp image editor lang file updates

MuchQuak committed:
remove notes

MuchQuak committed:
Change en translation of tpeditor to use multimedia instead of media

MuchQuak committed:
Tpeditor updated langs

MuchQuak committed:
update imagelib/search translations

MuchQuak committed:
updating multimedia changes, removing unused langs tags, and fixng other translations for taxa/index.php

MuchQuak committed:
imgdetail lang updates

MuchQuak committed:
indx langfile updates

MuchQuak committed:
removing dead code in lang files occurrence editor

MuchQuak committed:
add has audio flag to search

MuchQuak committed:
updated dwc export to have audio files with correct values for ac core

MuchQuak committed:
updated import tool to only import image media

Currently we are not supporting audio imports this will likely change in
the future

MuchQuak committed:
add audio symbol to listdisplay when audio is present in occurrence data

MuchQuak committed:
swapping image root to media root variables and aliasing them in symbase so upgrade is easier

MuchQuak committed:
update occurrence individual lang files to remove multimedia from file open lang tags

MuchQuak committed:
update map search lang files to have audio option

MuchQuak committed:
add audio filter to map search

MuchQuak committed:
update lang files for download for media update

MuchQuak committed:
exporter lang update multimedia

MuchQuak committed:
add image and audio tags to records collections list

MuchQuak committed:
Image tab lang updates

MuchQuak committed:
imagelib index lang updates

MuchQuak committed:
contributors lang updates

MuchQuak committed:
site map multimedia update

MuchQuak committed:
collprofile translations

MuchQuak committed:
fixing header not being translated contributors

MuchQuak committed:
fix has_audio and has_image warnings

MuchQuak committed:
tpimageeditor generalizing lang for opening large image to large file

MuchQuak committed:
pulling update from Development for upcoming pr

MuchQuak committed:
Removing dead code from media

MuchQuak committed:
Refactor of UploadStrategy to StorageStrategy and move to using Media class for update

MuchQuak committed:
Add Renaming to media files to update function in media class

MuchQuak committed:
Add better error handling and misc todos media

MuchQuak committed:
clean up remaining todos

MuchQuak committed:
allow remote audio to work

MuchQuak committed:
Add Audio Media support for data import

MuchQuak committed:
fix issue with specproccessor using imgid instead of media_id

MuchQuak committed:
cleaning up imgid usage with media table

MuchQuak committed:
Importextended media lang update

MuchQuak committed:
Updating Darwin Core Archive Publishing lang files to use media lang

MuchQuak committed:
Occurrence Editor Media Lang file update

MuchQuak committed:
Merge branch 'Development' into image-to-media

MuchQuak committed:
fixed other data multimedia import

MuchQuak committed:
Add status signal and occid linking to multimedia url mapper

MuchQuak committed:
move Allowed media mime types to symbase so it is standardized about what we support

MuchQuak committed:
fix backwards compat of mysqli_execute_query

MuchQuak committed:
Fix backwards compat by changing database trait to class

MuchQuak committed:
Remove enums from media.php for backwards compat. Replace them with class enum facades

MuchQuak committed:
add schema patch for media table

MuchQuak committed:
add error redundency

MuchQuak committed:
OccurrenceEditorImages Remove from occurrence editor. All functionality was moved to media

MuchQuak committed:
internationalize media errors

MuchQuak committed:
add drop table for deprecated_media to fix key issue

themerekat commented:
I've done some more research into using "media" versus "multimedia", and I think we need to consistently use the former rather than the latter.
![image](https://github.com/user-attachments/assets/eca84ae7-279e-49a7-b95b-957e639c1938)
https://dictionary.cambridge.org/us/dictionary/english/media

Multimedia seems to be most appropriate for when different types of media are mixed together into one presentation, which _kind of_ fits our search here, but perhaps just consistent use of "media" would be less confusing overall.
![image](https://github.com/user-attachments/assets/2d51c13f-d504-4109-b681-01656046fa04)
https://www.merriam-webster.com/dictionary/multimedia

@MuchQuak I know this has been an annoying back and forth, so I went ahead and changed them all myself 😅 See commit below. I did NOT change instances in the vendor files, classes, or Darwin Core Archive parser.

themerekat committed:
Change instances of "multimedia" to "media

In Spanish, largely "medios"

themerekat committed:
Misc lang file changes

themerekat commented:
In importextended.php, when you're uploading a media file, the dropdown list still contains "photographer" and "photographeruid" instead of "creator" and "creatoruid"
![image](https://github.com/user-attachments/assets/87b0e590-ea9c-42f7-94a6-b76227248444)

themerekat commented:
I'm trying to upload the attached file using importextended, and it doesn't seem to work.
[example_sounds.csv](https://github.com/user-attachments/files/17344430/example_sounds.csv)

Could it be because the links are to mp4 files, or is something else amiss here?

Furthermore, what's up with the "Empty dateIdentified values have been set as" message? I think this is leftover from when the type of import is a Determinations import.

![image](https://github.com/user-attachments/assets/c0104822-f314-4bfa-a246-2c1300fe26d1)

egbot committed:
Schema modifications

- Change naming of new fields from snake_case to camelCase
- Change input variables from snake_case to lowercase to match format of other variables associated with page

egbot committed:
First pass of updating media_id to mediaID

egbot committed:
Schema mods continued

- Remap media_id to mediaID

egbot committed:
Media Field Mapping continued

- Remap imgid foreign key to mediaID for tables imagekeywords, imagetag, specprocessorrawlabels, tmattributes
- Remap media API model and controller
- Misc remapping previously missed

MuchQuak committed:
Merge pull request #1945 from BioKIC/image-to-media-mods-egbot

Image to media mods egbot

MuchQuak committed:
Merge branch 'Development' into image-to-media

MuchQuak committed:
Adjust fk swaps to drop only if keys exist

MuchQuak committed:
fix query util

MuchQuak committed:
fix merge issue with query util

MuchQuak committed:
Merge up query util changes

MuchQuak committed:
clean up mysqli_execute_query reference

MuchQuak committed:
Fixing sql patch from merge

MuchQuak committed:
Pulling latest merge from development

MuchQuak committed:
fix overwritten functionality of allowing for empty params in executeQuery
